File: dashboard/src/signals.js
// VSS signal paths the dashboard renders — the UI-side mirror of the bridge
// catalog (bridge/signals.py). Path→UI-element mapping is inherently UI-side,
// so it lives here; the bridge stays the authority on which signals exist
// (the dashboard ignores any path it doesn't know how to draw).
//
// Keep these strings in sync with bridge/signals.py. If a path is renamed for a
// different VSS version, change it in both places (or fetch GET /signals).

export const P = {
  lowBeam: "Vehicle.Body.Lights.Beam.Low.IsOn",
  highBeam: "Vehicle.Body.Lights.Beam.High.IsOn",
  indLeft: "Vehicle.Body.Lights.DirectionIndicator.Left.IsSignaling",
  indRight: "Vehicle.Body.Lights.DirectionIndicator.Right.IsSignaling",
  hazard: "Vehicle.Body.Lights.Hazard.IsSignaling",
  lockDriver: "Vehicle.Cabin.Door.Row1.DriverSide.IsLocked",
  lockPass: "Vehicle.Cabin.Door.Row1.PassengerSide.IsLocked",
  doorDriverOpen: "Vehicle.Cabin.Door.Row1.DriverSide.IsOpen",
  windowDriver: "Vehicle.Cabin.Door.Row1.DriverSide.Window.Position",
  temp: "Vehicle.Cabin.HVAC.Station.Row1.Left.Temperature",
  ac: "Vehicle.Cabin.HVAC.IsAirConditioningActive",
  speed: "Vehicle.Speed",
  battery: "Vehicle.Powertrain.TractionBattery.StateOfCharge.Current",
};

// Short labels for the command feed / debugging.
export const LABELS = {
  [P.lowBeam]: "Low beam",
  [P.highBeam]: "High beam",
  [P.indLeft]: "Left indicator",
  [P.indRight]: "Right indicator",
  [P.hazard]: "Hazards",
  [P.lockDriver]: "Driver lock",
  [P.lockPass]: "Passenger lock",
  [P.doorDriverOpen]: "Driver door",
  [P.windowDriver]: "Driver window",
  [P.temp]: "Cabin temp",
  [P.ac]: "A/C",
  [P.speed]: "Speed",
  [P.battery]: "Battery",
};

export const labelFor = (path) => LABELS[path] || path;

// Truthy helper tolerant of bool / "true" / 1.
export const on = (v) => v === true || v === 1 || v === "true";

File: bridge/signals.py
"""VSS signal catalog for the demo dashboard.

These are COVESA VSS (Vehicle Signal Specification) paths. The Kuksa Databroker
loads a specific `vss.json`, and exact path names DIFFER between VSS versions
(e.g. older VSS used `Vehicle.Body.Lights.IsLowBeamOn` instead of
`Vehicle.Body.Lights.Beam.Low.IsOn`).

>>> Before wiring the real databroker, confirm these paths against the tree your
>>> databroker actually loaded:  `kuksa-client` -> getMetaData '*'  (or `getValue`)

Each entry: path -> {type, default}. `type` is informational (used for parsing
values coming in over HTTP/WebSocket as strings).
"""

# fmt: off
SIGNALS = {
    # Lighting
    "Vehicle.Body.Lights.Beam.Low.IsOn":                          {"type": "bool",  "default": False},
    "Vehicle.Body.Lights.Beam.High.IsOn":                         {"type": "bool",  "default": False},
    "Vehicle.Body.Lights.DirectionIndicator.Left.IsSignaling":    {"type": "bool",  "default": False},
    "Vehicle.Body.Lights.DirectionIndicator.Right.IsSignaling":   {"type": "bool",  "default": False},
    "Vehicle.Body.Lights.Hazard.IsSignaling":                     {"type": "bool",  "default": False},

    # Doors / locks / windows
    "Vehicle.Cabin.Door.Row1.DriverSide.IsLocked":                {"type": "bool",  "default": True},
    "Vehicle.Cabin.Door.Row1.PassengerSide.IsLocked":             {"type": "bool",  "default": True},
    "Vehicle.Cabin.Door.Row1.DriverSide.IsOpen":                  {"type": "bool",  "default": False},
    "Vehicle.Cabin.Door.Row1.DriverSide.Window.Position":         {"type": "uint8", "default": 0},   # 0 closed .. 100 open

    # Climate
    "Vehicle.Cabin.HVAC.Station.Row1.Left.Temperature":           {"type": "float", "default": 21.0},
    "Vehicle.Cabin.HVAC.IsAirConditioningActive":                 {"type": "bool",  "default": False},

    # Drivetrain / energy
    "Vehicle.Speed":                                              {"type": "float", "default": 0.0},
    "Vehicle.Powertrain.TractionBattery.StateOfCharge.Current":   {"type": "float", "default": 78.0},
}
# fmt: on

PATHS = list(SIGNALS.keys())


def default_state() -> dict:
    """Initial value for every signal."""
    return {path: meta["default"] for path, meta in SIGNALS.items()}


def coerce(path: str, value):
    """Best-effort coerce an incoming value to the signal's declared type."""
    meta = SIGNALS.get(path)
    if meta is None:
        return value
    t = meta["type"]
    try:
        if t == "bool":
            if isinstance(value, str):
                return value.strip().lower() in ("1", "true", "yes", "on")
            return bool(value)
        if t in ("uint8", "int"):
            return int(float(value))
        if t == "float":
            return float(value)
    except (TypeError, ValueError):
        return value
    return value
